PrefectHQ/prefect
# Resolving block references in prefect.yaml and adding keypath feature for blocks

Fixes #10388 and closes #10930.

Block references are now being successfully resolved in `prefect.yaml` outside of the `build/push/pull` sections.
As a dict or list parsed content of blocks can now be referenced via a 'keypath'.

Tested with added unit tests and example.

### Example

**Minimal example:**

Content of `work-pools` JSON Block:
```json
{
  "ecs": {
     "name": "ecs-work-pool",
   },
   "ny4": {
     "names": [
       "ny4-first-pool",
       "ny4-second-pool",
    ]
  }
}
```

Value for `default-work-queue-name` String Block:
```text
default-work-queue
```

`prefect.yaml` (slight deviation from #10388 to demonstrate keypath usage):
```yaml
definitions:
    work_pools:
        ecs_work_pool: &ecs_work_pool
            name: "{{ prefect.blocks.json.work-pools.value.ecs.name }}"
            work_queue_name: "{{ prefect.blocks.string.default-work-queue-name }}"
            job_variables:
                image: "*****.dkr.ecr.us-east-2.amazonaws.com/py310-base:0.0.3"

        ny4_work_pool: &ny4_work_pool
            name: "{{ prefect.blocks.json.work-pools.value.ny4.names[1] }}"
            work_queue_name: default
            job_variables:
                image: "****.dkr.ecr.us-east-2.amazonaws.com/py310-base:0.0.3"


# the deployments section allows you to provide configuration for deploying flows
deployments:
- name: test
  version: 0.1
  tags: []
  description: "test flow"
  schedule: {}
  entrypoint: "flow.py:test_flow"
  parameters: {}
  work_pool: *ecs_work_pool
- name: another-test
  version: 0.1
  tags: []
  description: "test flow"
  schedule: {}
  entrypoint: "flow.py:test_flow"
  parameters: {}
  work_pool: *ny4_work_pool
```

`flow.py`:
```python
from prefect import flow

@flow
def test_flow():
    print("I am a very good test")
```

Following these steps you can check out the result:

1. Create a new directory and copy `flow.py` and `prefect.yaml` files in there.
2. Have a Server/Cloud Workspace somewhere running and make sure your prefect config points to it.
3. Create the two blocks with the specified content with the UI.
4. Create the work pools `ecs-work-pool` and `ny4-second-pool`. (Can this be done via the API btw?)
5. Run `prefect deploy --all`.
6. You should see in the UI that everything was resolved correctly or/and run `prefect deployment inspect test-flow/test`/`prefect deployment inspect test-flow/another-test`

**Some additional info about the keypath feature**:

Take the base files from above to try it out yourself.

- If a path cannot be resolved an error is raised:

> ```yaml
> definitions:
>     work_pools:
>         ecs_work_pool: &ecs_work_pool
>             name: "{{ prefect.blocks.json.work-pools.value.ecs.this-key-does-not-exist }}"  # <----
>             work_queue_name: "{{ prefect.blocks.string.default-work-queue-name }}"
>             job_variables:
>                 image: "*****.dkr.ecr.us-east-2.amazonaws.com/py310-base:0.0.3"
>                 
>                 
> #Output
> Invalid template: '{{ prefect.blocks.json.work-pools.value.ecs.this-key-does-not-exist }}'. Found 'value' keyword with additional keys in template, but could not resolve the keypath in the block 
> document data.
> ```

- If a block is referenced whose content is not parsed as a dict or list internally an error is raised:

> ```yaml
> definitions:
>     work_pools:
>         ecs_work_pool: &ecs_work_pool
>             name: "{{ prefect.blocks.string.default-work-queue-name.value.not-a-dict }}"  # <----
>             work_queue_name: "{{ prefect.blocks.string.default-work-queue-name }}"
>             job_variables:
>                 image: "*****.dkr.ecr.us-east-2.amazonaws.com/py310-base:0.0.3"
>                 
> # Output:
> Invalid template: '{{ prefect.blocks.string.default-work-queue-name.value.not-a-dict }}'. Found 'value' keyword with additional keys in reference, but the content of the block was not parsed as a 
> dict or list.
> ```

- You can traverse lists and nested keys. Take a look at `tests/utilities/test_templating:TestResolveBlockDocumentReferences` for that.

Hope you like it, looking forward to your feedback! 😊 

### Checklist


	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.
  

For documentation changes:

- [ ] This pull request includes redirect settings in `netlify.toml` for files that are removed or renamed
